influxdb3_server/src/http.rs

```diff
@@ -1491,70 +1493,173 @@ async fn record_batch_stream_to_body(
+                                Err(e) => return Poll::Ready(Some(Err(e))),
+                            };
+                            let mut writer = if self.first_poll {
+                                self.first_poll = false;
+                                arrow_csv::Writer::new(Vec::new())
+                            } else {
+                                arrow_csv::WriterBuilder::new()
+                                    .with_header(false)
+                                    .build(Vec::new())
+                            };
+                            writer.write(&batch).unwrap();
+                            Poll::Ready(Some(Ok(Bytes::from(writer.into_inner()))))
+                        }
+                        Poll::Ready(None) => Poll::Ready(None),
+                        Poll::Pending => Poll::Pending,
+                    }
+                }
+            }
+
+            let mut future = CsvFuture {
+                first_poll: true,
+                stream,
+            };
+            Ok(Body::wrap_stream(futures::stream::poll_fn(move |ctx| {
+                future.poll_unpin(ctx)
+            })))
         }
         QueryFormat::Json => {
-            let batches = stream.try_collect::<Vec<RecordBatch>>().await?;
-            to_json(batches).map(Body::from)
+            struct JsonFuture {
+                state: State,
+                stream: Pin<Box<dyn RecordBatchStream + Send>>,
+            }
+
+            enum State {
+                FirstPoll,
+                Body,
+                Done,
+            }
+
+            #[derive(Default, Debug)]
+            struct JsonMiddle;
+
+            impl arrow_json::writer::JsonFormat for JsonMiddle {
+                fn start_row<W: std::io::Write>(
+                    &self,
+                    writer: &mut W,
+                    is_first_row: bool,
+                ) -> std::result::Result<(), arrow_schema::ArrowError> {
+                    if !is_first_row {
+                        writer.write_all(b",")?;
+                    }
+
+                    Ok(())
+                }
+            }
+
+            impl Future for JsonFuture {
+                type Output = Option<Result<Bytes, DataFusionError>>;
+
+                fn poll(
+                    mut self: Pin<&mut Self>,
+                    ctx: &mut std::task::Context<'_>,
+                ) -> Poll<Self::Output> {
+                    match dbg!(self.stream.poll_next_unpin(ctx)) {
+                        Poll::Ready(Some(batch)) => {
+                            let batch = match batch {
+                                Ok(batch) => batch,
+                                Err(e) => return Poll::Ready(Some(Err(e))),
+                            };
+                            match self.state {
+                                State::FirstPoll => {
+                                    let mut writer = arrow_json::ArrayWriter::new(Vec::new());
```

Comment:
Would it be possible (or make sense if it is possible) to store the `ArrayWriter` in the state enum so it doesn't have to be reconstructed for every `RecordBatch`? Also, does the `ArrayWriter` automatically take care of adding the opening square bracket? Just curious because I see the closing square bracket written in this future but not the opening one when the stream is non-empty.

Reply:
Oh I was just thinking about this and realized it's because you're just using the writer for its formatting capabilities and immediately converting to bytes as the return value of this iteration :facepalm: